dg.go

```diff
@@ -223,25 +223,43 @@ func (d *directedGraph[NodeType]) PushStartingNodes() error {
 	defer d.lock.Unlock()
 
 	for nodeID, n := range d.nodes {
+		// Search for nodes that have either no outstanding dependencies, or only
+		// obviated dependencies.
 		if len(n.outstandingDependencies) == 0 {
 			d.readyForProcessing[nodeID] = n
+		} else {
+			// Has nodes left. Check to see if they are all obviated.
+			hasNonObviatedDependencies := false
+			for _, dependency := range n.outstandingDependencies {
+				if dependency != ObviatedDependency {
+					hasNonObviatedDependencies = true
+					break
+				}
+			}
+			if !hasNonObviatedDependencies {
+				d.readyForProcessing[nodeID] = n
+			}
 		}
 	}
```

Comment:
Alternatively,
```go
	for nodeID, n := range d.nodes {
		isReady := true
		for _, dependency := range n.outstandingDependencies {
			if dependency != ObviatedDependency {
				isReady = false
				break
			}
		}
		if isReady {
			d.readyForProcessing[nodeID] = n
		}
	}
```

That is, there is no particular need to check `len(n.outstandingDependencies) == 0`, because, without that check, you get the same effect with less, simpler code (that is, if there are no outstanding dependencies, then control doesn't enter the inner loop, it finds the node to be ready, and adds it to the list.

(And, I changed `!hasNonObviatedDependencies` to `isReady`, which I think is easier to read and understand.)

But, if you're up for it, I think we can simplify this further, using a [`continue` statement with a labeled loop](https://go.dev/ref/spec#Continue_statements):
```go
nextNode:
	for nodeID, n := range d.nodes {
		for _, dependency := range n.outstandingDependencies {
			if dependency != ObviatedDependency {
				continue nextNode
			}
		}
		d.readyForProcessing[nodeID] = n
	}
```

Reply:
Sure. I'll go with the continue version.